[Test]: Enable unit tests to run on non-CUDA environment

hickeyma opened the pull request:
Enable unit tests where applicable to run on non-CUDA environment. The goal is to be able to run `pytest` and only the relevant tests run.

Unit tests:

- [x] Tests for storage backend: `pytest tests/v1/storage_backend/`
- [x] Skip NIXL tests if nixl not installed: `pytest tests/disagg tests/v1/test_nixl_storage.py`
- [x] Benchmark tests: `pytest tests/benchmarks/` (caveat: skipping `benchmark` tests for now due to CUDA complexity)
- [x] Tests in `tests`: `pytest tests/*.py` (caveat: skipping `tests/test_serde.py` for now due to CUDA complexity)
- [x] Tests in `tests/v1`: `pytest tests/v1*.py`

CI:

- [x] Add GH workflow to test the unit tests as non-CUDA

FIX #1503 

**Note:** 

- Used Mac M-series as the test system
- Changes made to core code was in a clean incisive fashion with minimal of change to enable some of the tests to run on a non-CUDA environment. There are not meant to enable LMCache to be able to run fully on non-CUDA envoronments.
- Opened PR #1664 for follow on work where finding an equivalent non CUDA implementation is more complex
- There are a numbers of warnings listed post unit test run. The warnings are inherited and not caused by this PR (https://buildkite.com/lmcache/unit-tests/builds/5196#0199516d-ea47-4f7a-b013-d578a66297b2)

hickeyma committed:
Enable storage backend unit tests to run as non-CUDA

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@ApostaC @maobaolong @sammshen @YaoJiayi  First go at this by enabling the storage backend tests to run.

@YaoJiayi Looking for patterns on how we handle the C++ CUDA enhancements as Python non-CUDA equivalents. This is a first go but will need work. I wanted to get it started so we can build on it.

ApostaC commented:
Thanks @hickeyma for leading this effort! Will take a look detailed look!

maobaolong commented:
@hickeyma Thank you advance for this work!

I'm trying to run `pytest tests/v1/storage_backend/` , but I failed, the following is part of the output.

<img width="1610" height="1576" alt="image" src="https://github.com/user-attachments/assets/97aaa952-68af-45b4-9c33-022fb563b152" />

BTW, I use macbook pro M3.

hickeyma committed:
Add pytest asyncio dependency for async tests

Tests in tests/v1/storage_backend/test_gds_backend.py were not
being run as they required async library.

This adds the pytest asyncio package which enabkles the tests to be run
by default rather than skipping them.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
> I'm trying to run pytest tests/v1/storage_backend/ , but I failed, the following is part of the output.

@maobaolong I am unable to reproduce on a Mac M3 also. Not sure the differences between our devices.

Checking out the error it says the following:

_"Pinned (page‑locked) memory is a CPU‑side optimization for faster, asynchronous transfers to a CUDA device. If there’s no CUDA context or no allocator registered, PyTorch can’t actually pin the memory, so it complains."_ 

This doesn't explain why it works for me though.

hickeyma committed:
Add unit test GitHub workflow for non-CUDA tests

This workflow will run non-CUDA unit tests which can run on the
standard runners which are not GPU enabled.

This could help catch issues prior to the running the full tests
on GPU systems. This could be helpful towards resource savings.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Pinned memory does not work on non-CUDA systems

Pinned (page‑locked) memory is a CPU‑side optimization for faster,
asynchronous transfers to a CUDA device. If there’s no CUDA context or
no allocator registered, PyTorch can’t actually pin the memory, so it complains.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

maobaolong commented:
@hickeyma Thanks for your reply, but I'm curious about what device you are using now? Apple mackbook pro M1/2/3/4? After a double check, my  device is `Apple M4 Pro`.

I wonder why it can work on your device?

hickeyma committed:
Disable tests for Python 3.9 until supported added to project for it

Ref: https://github.com/LMCache/LMCache/pull/1584

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
>@hickeyma Thanks for your reply, but I'm curious about what device you are using now? Apple mackbook pro M1/2/3/4? After a double check, my device is Apple M4 Pro.
I wonder why it can work on your device?

@maobaolong It is a M3. I am unsure why it runs on my system. There maybe something that I enabled to get something to work previously that is allowing the test to run. I'll look into it later to better understand it.

In the meantime, the tests did failed on the GH runner: https://github.com/LMCache/LMCache/actions/runs/17680597527/job/50253625508> I have therefore disabled the `pin_memory` as suggested in https://github.com/LMCache/LMCache/pull/1575#issuecomment-3285498184.

The unit tests are no passing in CI.

hickeyma committed:
Fix memory issues in non-CUDA memory  allocator functions

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on lmcache/utils.py:
```diff
@@ -369,3 +369,26 @@ def mock_up_broadcast_fn(t: torch.Tensor, i: int) -> None:
+    tensor = torch.empty(size, dtype=torch.uint8, pin_memory=True)
+
+    # Return address of first element of tensor
+    return tensor.data_ptr()
```
Good suggestions.

hickeyma commented on tests/v1/storage_backend/test_local_disk_backend.py:
```diff
@@ -195,6 +195,10 @@ def test_get_blocking_key_not_exists(self, local_disk_backend):
+    @pytest.mark.skipif(
+        not torch.cuda.is_available(),
+        reason="Requires CUDA for LocalDiskBackend load_bytes_from_disk",
+    )
```
Removed skip

hickeyma commented on tests/v1/storage_backend/test_local_disk_backend.py:
```diff
@@ -220,6 +224,10 @@ def test_async_load_bytes_from_disk(self, local_disk_backend):
+    @pytest.mark.skipif(
+        not torch.cuda.is_available(),
+        reason="Requires CUDA for LocalDiskBackend load_bytes_from_disk",
+    )
```
Removed skip

maobaolong commented:
@hickeyma Glad to sync that I passed the tests `pytest tests/v1/storage_backend/` by the new state of this PR, commit hash is [60881bb].

```shell
(base) ~/projects/LMCache git:[60881bb]
pytest tests/v1/storage_backend/
============================================== test session starts ==============================================
platform darwin -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0
rootdir: /Users/msy/projects/LMCache/tests
configfile: pytest.ini
plugins: anyio-4.8.0
collected 46 items                                                                                              

tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_init SKIPPED (TestGdsBackend runs ...) [  2%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_str SKIPPED (TestGdsBackend runs o...) [  4%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_key_to_path_and_insert_key SKIPPED     [  6%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_contains_key_not_exists SKIPPED (T...) [  8%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_contains_key_exists SKIPPED (TestG...) [ 10%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_exists_in_put_tasks SKIPPED (TestG...) [ 13%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_submit_put_task_and_get_blocking SKIPPED [ 15%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_batched_submit_put_task SKIPPED (T...) [ 17%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_get_blocking_key_not_exists SKIPPED    [ 19%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_close SKIPPED (TestGdsBackend runs...) [ 21%]
tests/v1/storage_backend/test_gds_backend.py::TestGdsBackend::test_pin_unpin_not_implemented SKIPPED      [ 23%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_init PASSED                 [ 26%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_init_with_lookup_server_and_worker PASSED [ 28%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_init_with_layerwise_config PASSED [ 30%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_str PASSED                  [ 32%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_contains_key_not_exists PASSED [ 34%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_contains_key_exists PASSED  [ 36%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_exists_in_put_tasks PASSED  [ 39%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_submit_put_task PASSED      [ 41%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_submit_put_task_reinsert PASSED [ 43%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_batched_submit_put_task PASSED [ 45%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_batched_submit_put_task_disabled PASSED [ 47%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_get_blocking_key_not_exists PASSED [ 50%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_get_blocking_key_exists PASSED [ 52%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_pin_unpin PASSED            [ 54%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_remove PASSED               [ 56%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_remove_non_existent PASSED  [ 58%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_remove_with_worker PASSED   [ 60%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_allocate PASSED             [ 63%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_allocate_with_format PASSED [ 65%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_allocate_with_layerwise_config PASSED [ 67%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_batched_allocate PASSED     [ 69%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_get_keys PASSED             [ 71%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_get_keys_empty PASSED       [ 73%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_concurrent_access PASSED    [ 76%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_thread_safety PASSED        [ 78%]
tests/v1/storage_backend/test_local_cpu_backend.py::TestLocalCPUBackend::test_ref_count_management PASSED [ 80%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init PASSED               [ 82%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init_with_lookup_server_and_worker PASSED [ 84%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_str PASSED                [ 86%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_key_to_path PASSED        [ 89%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_contains_key_not_exists PASSED [ 91%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_get_blocking_key_not_exists PASSED [ 93%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_async_load_bytes_from_disk PASSED [ 95%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_load_bytes_from_disk PASSED [ 97%]
tests/v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_file_operations_error_handling PASSED [100%]

=============================================== warnings summary ================================================
tests/v1/storage_backend/test_gds_backend.py:127
  /Users/msy/projects/LMCache/tests/v1/storage_backend/test_gds_backend.py:127: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

tests/v1/storage_backend/test_gds_backend.py:148
  /Users/msy/projects/LMCache/tests/v1/storage_backend/test_gds_backend.py:148: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init_with_lookup_server_and_worker
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_str
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_key_to_path
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_contains_key_not_exists
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_get_blocking_key_not_exists
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_async_load_bytes_from_disk
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_load_bytes_from_disk
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_file_operations_error_handling
  /opt/homebrew/Caskroom/miniconda/base/lib/python3.12/asyncio/base_events.py:716: RuntimeWarning: coroutine 'AsyncPQThreadPoolExecutor._worker' was never awaited
    self._ready.clear()
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================== 35 passed, 11 skipped, 11 warnings in 0.69s ==================================
```

hickeyma commented:
There is an issue with the v1 GDS backend tests as they will not run in the Buildkite test environment. I opened #1592 to cover the issue. I am therefore going to disable the GDS backend tests (0565b33911fb9ff912b8fe324508d45a1e79e0f7) for now so that CI can pass.

hickeyma committed:
Revert "Add pytest asyncio dependency for async tests"

This reverts commit 697554a98657eba65449ef95272e0fff5974d551.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Disable GDS storage backend tests

Tests currently don't run on test system. Issue opened:
https://github.com/LMCache/LMCache/issues/1592

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Revert "Revert "Add pytest asyncio dependency for async tests""

This reverts commit 706156b9d2fc5b36b1252aeb8c3bbc2295fd3a9e.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add check to nixl tests that nixl is imported

Skip NIXL tests is the nixl package is not installed

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Enable non-CUDA unit testing in tests root directory files

Files in /tests

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

maobaolong commented:
@hickeyma Look forward to merge this PR! It really make it easy to develop UTs for backends or remote connectors. 😄.

The PR https://github.com/LMCache/LMCache/pull/1594 is verified on `Macbook Pro M4` locally powered by pre-merge this PR.

hickeyma committed:
Disable benchmark tests as no non cuda solution yet

Code tested by benchmark test is very CUDA specific and no
solution at this time. Therefore, skip tests if not CIUDA.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Enable v1 unit tests

Tests in `tests/v1`

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
> @hickeyma Look forward to merge this PR! It really make it easy to develop UTs for backends or remote connectors. 😄.
The PR https://github.com/LMCache/LMCache/pull/1594 is verified on Macbook Pro M4 locally powered by pre-merge this PR.

There you go @maobaolong. I would appreciate if you could try it out fully locally (i.e. `pytest`) and would welcome your review comments.

ApostaC commented on lmcache/utils.py:
```diff
@@ -384,3 +384,26 @@ def mock_up_broadcast_fn(t: torch.Tensor, i: int) -> None:
+    # First-touch initialization (forces physical allocation)
+    tensor.fill_(0.0)
+
+    return tensor
```
I like this idea of having the same set of C functions implemented in Python for non-CUDA platforms.

Can we separate those to a new source file so that we can import it as:
``` 
if torch.cuda.is_available():
    import lmcache.c_ops as lmc_ops
else:
    import lmcache.mockup_ops as lmc_ops  # could have different names 
```

In this way, we can avoid having if conditions in the runtime code (like in `memory_management.py`).

hickeyma replied:
Good idea. I moved them top their own file: `lmcache/non_cuda_equivalents.py`.

> In this way, we can avoid having if conditions in the runtime code (like in memory_management.py).

Unfortunately, it won't be possible to remove the conditionals. This is because of how the memory is allocated by the C/CPP CUDA functions and how the non-CUDA Python equivalents.

C function allocates the memory and return a pointer. This memory will remain until explicitly de-alloacted which is done in `close()` method in `memory_management.py`.  An `array` type is assigned to the pointer which is then used to create a `tensor` object which is assigned to the `buffer` property name.

However, with the equivalent Python function, a tensor objects is instantiated and needs to be allocated prior to returning from the function. It then needs to be assigned to a class property to keep reference to the object. It is then assigned directly to the `buffer` property name.

They are essentially different means thats are not common.

sammshen replied:
@hickeyma how about hacking the mock interface to return and manage raw pointers as well? 

maybe something like `lmcache/non_cuda_equivalents.py`:

```python

# SPDX-License-Identifier: Apache-2.0
#
# Non-CUDA fallback implementations for CUDA-specific operations.
# These mimic the CUDA C API surface by returning raw pointers.
#
# Third Party
import torch
import ctypes

# Global registry to keep tensors alive (avoid GC).
_tensor_registry: dict[int, torch.Tensor] = {}


def alloc_pinned_ptr(size: int, device_id: int = 0) -> int:
    """Allocate pinned memory and return a raw pointer (int)."""
    # Create CPU tensor (not actually pinned, since CUDA isn't available).
    tensor = torch.empty(size, dtype=torch.uint8, pin_memory=False)

    # First-touch initialization (forces physical allocation).
    tensor.fill_(0)

    # Store in registry to prevent GC.
    ptr = tensor.data_ptr()
    _tensor_registry[ptr] = tensor

    return ptr


def free_pinned_ptr(ptr: int) -> None:
    """Free a previously allocated pinned pointer."""
    _tensor_registry.pop(ptr, None)


def alloc_pinned_numa_ptr(size: int, numa_id: int = 0) -> int:
    """Allocate pinned memory with NUMA awareness (mocked)."""
    # For non-CUDA, we ignore numa_id and just allocate normally.
    tensor = torch.empty(size, dtype=torch.uint8, pin_memory=False)
    tensor.fill_(0)

    ptr = tensor.data_ptr()
    _tensor_registry[ptr] = tensor

    return ptr


def free_pinned_numa_ptr(ptr: int, size: int | None = None) -> None:
    """Free a previously allocated NUMA pointer (mocked)."""
    _tensor_registry.pop(ptr, None)
    
    
```

hickeyma replied:
Thanks @sammshen for the great suggestion and the code to handle the pointer within the function. Added the code in 37ca0d813886fac2019b879aea0310b4c907ff8f.

ApostaC reviewed (commented):
Thanks for the great effort! Did a quick pass, and will do another round of review later

hickeyma committed:
Move the fallback non-CUDA implementation to their own file

Review comment:
- https://github.com/LMCache/LMCache/pull/1575#discussion_r2361460649

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Change non-CUDA equivalents functions to return pointer to tensor object

This means the functions have same signature as their CUDA equivalents
and can therefore be neabled at import time and remove need for checking
for CUDA throughout the code thereafter.

Thank you to Samuel for suggestion means to handle the pointer from within
the functions to be equivalent.

Co-authored-by: Samuel Shen <slshen@uchicago.edu>
Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sammshen reviewed (approved):
LGTM!

hickeyma committed:
Fix formatting

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@maobaolong @ApostaC Do you mind doing a review again? I think it is ready to go and it would be good to get the capability merged. Also, the rebasing is getting challenging! :)

maobaolong commented:
@hickeyma I run `pytest tests/v1/storage_backend/` got 2 failed tests.
<details>
```console

==================================================== ERRORS =====================================================
____________________________ ERROR at setup of TestCacheGenSerde.test_round_trip_cpu ____________________________

self = <test_cachegen_serde.TestCacheGenSerde object at 0x16d2b5b50>

    @pytest.fixture
    def serde(self):
        config = create_test_config()
        metadata = create_test_metadata()
>       serializer, deserializer = CreateSerde("cachegen", metadata, config)

tests/v1/storage_backend/test_cachegen_serde.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lmcache/v1/storage_backend/naive_serde/__init__.py:35: in CreateSerde
    CacheGenSerializer(config, metadata),
lmcache/v1/storage_backend/naive_serde/cachegen_encoder.py:23: in __init__
    self.key_bins = self.make_key_bins(self.cachegen_config)
lmcache/v1/storage_backend/naive_serde/cachegen_encoder.py:32: in make_key_bins
    return ret.cuda()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def _lazy_init():
        global _initialized, _queued_calls
        if is_initialized() or hasattr(_tls, "is_initializing"):
            return
        with _initialization_lock:
            # We be double-checked locking, boys!  This is OK because
            # the above test was GIL protected anyway.  The inner test
            # is for when a thread blocked on some other thread which was
            # doing the initialization; when they get the lock, they will
            # find there is nothing left to do.
            if is_initialized():
                return
            # It is important to prevent other threads from entering _lazy_init
            # immediately, while we are still guaranteed to have the GIL, because some
            # of the C calls we make below will release the GIL
            if _is_in_bad_fork():
                raise RuntimeError(
                    "Cannot re-initialize CUDA in forked subprocess. To use CUDA with "
                    "multiprocessing, you must use the 'spawn' start method"
                )
            if not hasattr(torch._C, "_cuda_getDeviceCount"):
>               raise AssertionError("Torch not compiled with CUDA enabled")
E               AssertionError: Torch not compiled with CUDA enabled

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/torch/cuda/__init__.py:310: AssertionError
___________________________ ERROR at setup of TestCacheGenSerde.test_device_handling ____________________________

self = <test_cachegen_serde.TestCacheGenSerde object at 0x16d2b56d0>

    @pytest.fixture
    def serde(self):
        config = create_test_config()
        metadata = create_test_metadata()
>       serializer, deserializer = CreateSerde("cachegen", metadata, config)

tests/v1/storage_backend/test_cachegen_serde.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lmcache/v1/storage_backend/naive_serde/__init__.py:35: in CreateSerde
    CacheGenSerializer(config, metadata),
lmcache/v1/storage_backend/naive_serde/cachegen_encoder.py:23: in __init__
    self.key_bins = self.make_key_bins(self.cachegen_config)
lmcache/v1/storage_backend/naive_serde/cachegen_encoder.py:32: in make_key_bins
    return ret.cuda()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def _lazy_init():
        global _initialized, _queued_calls
        if is_initialized() or hasattr(_tls, "is_initializing"):
            return
        with _initialization_lock:
            # We be double-checked locking, boys!  This is OK because
            # the above test was GIL protected anyway.  The inner test
            # is for when a thread blocked on some other thread which was
            # doing the initialization; when they get the lock, they will
            # find there is nothing left to do.
            if is_initialized():
                return
            # It is important to prevent other threads from entering _lazy_init
            # immediately, while we are still guaranteed to have the GIL, because some
            # of the C calls we make below will release the GIL
            if _is_in_bad_fork():
                raise RuntimeError(
                    "Cannot re-initialize CUDA in forked subprocess. To use CUDA with "
                    "multiprocessing, you must use the 'spawn' start method"
                )
            if not hasattr(torch._C, "_cuda_getDeviceCount"):
>               raise AssertionError("Torch not compiled with CUDA enabled")
E               AssertionError: Torch not compiled with CUDA enabled

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/torch/cuda/__init__.py:310: AssertionError
=============================================== warnings summary ================================================
tests/v1/storage_backend/test_gds_backend.py:128
  /Users/msy/projects/LMCache/tests/v1/storage_backend/test_gds_backend.py:128: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

tests/v1/storage_backend/test_gds_backend.py:149
  /Users/msy/projects/LMCache/tests/v1/storage_backend/test_gds_backend.py:149: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_init_with_lookup_server_and_worker
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_str
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_key_to_path
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_contains_key_not_exists
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_get_blocking_key_not_exists
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_async_load_bytes_from_disk
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_load_bytes_from_disk
v1/storage_backend/test_local_disk_backend.py::TestLocalDiskBackend::test_file_operations_error_handling
  /opt/homebrew/Caskroom/miniconda/base/lib/python3.12/asyncio/base_events.py:716: RuntimeWarning: coroutine 'AsyncPQThreadPoolExecutor._worker' was never awaited
    self._ready.clear()
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================ short test summary info ============================================
ERROR tests/v1/storage_backend/test_cachegen_serde.py::TestCacheGenSerde::test_round_trip_cpu - AssertionError: Torch not compiled with CUDA enabled
ERROR tests/v1/storage_backend/test_cachegen_serde.py::TestCacheGenSerde::test_device_handling - AssertionError: Torch not compiled with CUDA enabled
============================= 42 passed, 11 skipped, 11 warnings, 2 errors in 1.12s 
```
</details>

maobaolong commented:
I run it on my laptop which is Macos M4

hickeyma commented:
@maobaolong The test failing seems to be `tests/v1/storage_backend/test_cachegen_serde.py`. Is this a new test as I cannot see it in the codebase?

maobaolong commented:
> @maobaolong The test failing seems to be `tests/v1/storage_backend/test_cachegen_serde.py`. Is this a new test as I cannot see it in the codebase?

Appreciate for figure out the extra test file for me, It can all passed after removed it.

maobaolong commented:
@hickeyma I've verify the following.
- [x] pytest tests/v1/storage_backend/ `42 passed, 11 skipped, 11 warnings in 0.78s`
- [x] pytest tests/v1/ `101 passed, 65 skipped, 14 warnings in 32.61s`
- [x] pytest tests/  `128 passed, 101 skipped, 18 warnings in 54.80s `
- [x] pytest tests/benchmarks/ `15 skipped, 4 warnings in 0.01s`
- [x] pytest tests/disagg tests/v1/test_nixl_storage.py ` 6 skipped in 0.01s`

Thanks for this great work,  LGTM X 3

maobaolong commented:
@hickeyma Would you like to merge this PR now? You would suffer conflict and rebasing if we merge this too late.

hickeyma commented:
Thanks @ApostaC @maobaolong @sammshen for the reviews.

@ApostaC I merged the PR before anymore rebasing is required. If you have any open issues, I can handle them in a follow on PR.

maobaolong commented:
<img width="1206" height="1108" alt="Clipboard_Screens@ot_1758677015" src="https://github.com/user-attachments/assets/cf77a0c2-ef3e-4eca-a003-7f6b187c75b6" />


@hickeyma Why we need `self.stream` within `LocalCPUBackend` although `CPU doesn't need streams` ? Could you just remove `self.stream`?

hickeyma commented:
> @hickeyma Why we need self.stream within LocalCPUBackend although CPU doesn't need streams ? Could you just remove self.stream?

I suppose you could @maobaolong. I was just showing that it won't be set to anything for CPU. More for code readability so others know why its this way.

maobaolong commented:
https://github.com/LMCache/LMCache/pull/1673 @hickeyma submitted a PR to remove it just now.

hickeyma commented:
> https://github.com/LMCache/LMCache/pull/1673 @hickeyma submitted a PR to remove it just now.

Ah, see what you mean now. Missed that it was unused at the time when adding the if statement. Nice catch, thanks.
